### PR TITLE
http/soc: Various implementations and fixes

### PR DESCRIPTION
**Pull request by FearlessTobi**
Allows the N3DS ("SKATER") browser to successfully load Google.com on Citra.
Please review commit-by-commit.

Screenshot:
image

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6828)

### CODE REVIEW

File: src/core/CMakeLists.txt [hunk 1]
```diff
@@ -419,12 +419,12 @@ add_library(citra_core STATIC
...
-    hle/service/ssl_c.cpp
-    hle/service/ssl_c.h
-    hle/service/y2r_u.cpp
-    hle/service/y2r_u.h
+    hle/service/soc/soc_u.cpp
+    hle/service/soc/soc_u.h
+    hle/service/ssl/ssl_c.cpp
+    hle/service/ssl/ssl_c.h
+    hle/service/y2r/y2r_u.cpp
+    hle/service/y2r/y2r_u.h
```
Comment (Steveice10) [inline src/core/CMakeLists.txt:427]: nit: Y2R is part of the camera module so we may want to put this under the cam folder, similar to how APT and NS are together.

File: src/core/hle/service/http/http_c.cpp [hunk 2]
```diff
@@ -55,11 +55,31 @@ const ResultCode ERROR_WRONG_CERT_HANDLE = // 0xD8A0A0C9
 const ResultCode ERROR_CERT_ALREADY_SET = // 0xD8A0A03D
     ResultCode(61, ErrorModule::HTTP, ErrorSummary::InvalidState, ErrorLevel::Permanent);
 
+static std::pair<std::string, std::string> splitUrl(const std::string& url) {
+    const std::string prefix = "://";
+    const auto scheme = url.find(prefix);
+    const auto scheme_end = scheme == std::string::npos ? 0 : scheme + prefix.length();
```
Comment (Steveice10) [inline src/core/hle/service/http/http_c.cpp:61]: The variable naming here is a little confusing, for example "scheme" is actually the position of the end of the scheme.

File: src/core/hle/service/http/http_c.cpp [hunk 3]
```diff
@@ -285,6 +305,99 @@ void HTTP_C::BeginRequestAsync(Kernel::HLERequestContext& ctx) {
...
+            context_handle, *session_data->current_http_context);
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
+        rb.Push(ERROR_STATE_ERROR);
+        return;
+    }
+
+    auto itr = contexts.find(context_handle);
+    ASSERT(itr != contexts.end());
+
+    itr->second.request_future.wait_for(std::chrono::nanoseconds(timeout));
```
Comment (Steveice10) [inline src/core/hle/service/http/http_c.cpp:395]: Can we make these two receive command share common code, with the only difference being whether there's a timeout or not?

Also should there be an error returned on timeout?

File: src/core/hle/service/http/http_c.cpp [hunk 4]
```diff
@@ -549,6 +662,101 @@ void HTTP_C::AddPostDataAscii(Kernel::HLERequestContext& ctx) {
...
+    if (!session_data->current_http_context) {
+        LOG_ERROR(Service_HTTP, "Tried to make a request without a bound context");
+
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
+        rb.Push(ResultCode(ErrorDescription::NotImplemented, ErrorModule::HTTP,
+                           ErrorSummary::Internal, ErrorLevel::Permanent));
+        return;
+    }
+
+    if (session_data->current_http_context != context_handle) {
```
Comment (Steveice10) [inline src/core/hle/service/http/http_c.cpp:691]: We do these same three state checks a lot, might want to move them to a common state checking function.

File: src/core/hle/service/http/http_c.cpp [hunk 5]
```diff
@@ -549,6 +662,101 @@ void HTTP_C::AddPostDataAscii(Kernel::HLERequestContext& ctx) {
...
+            context_handle, *session_data->current_http_context);
+        IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
+        rb.Push(ERROR_STATE_ERROR);
+        return;
+    }
+
+    auto itr = contexts.find(context_handle);
+    ASSERT(itr != contexts.end());
+
+    itr->second.request_future.wait_for(std::chrono::nanoseconds(timeout));
```
Comment (Steveice10) [inline src/core/hle/service/http/http_c.cpp:752]: Ditto here for reusing code for the command given the only difference is a timeout.

File: src/core/hle/service/http/http_c.cpp (hunk 3 above)
Comment (FearlessTobi) [reply src/core/hle/service/http/http_c.cpp]: Yes, therre probably should but I have not figured out which code it is yet.
That's why I left it as stubbed.

**Review by Steveice10 — APPROVED** [review]

**Comment by Adamillo** [comment]
What about the old 3DS browser?

**Comment by FearlessTobi** [comment]
> What about the old 3DS browser?

I do not own an old 3ds,

**Comment by Adamillo** [comment]
I see. I could test the old3ds browser by downloading the old3ds specific files from Citra itself, since I have the proper keys

File: src/core/hle/service/http/http_c.cpp [hunk 6]
```diff
@@ -55,11 +55,31 @@ const ResultCode ERROR_WRONG_CERT_HANDLE = // 0xD8A0A0C9
 const ResultCode ERROR_CERT_ALREADY_SET = // 0xD8A0A03D
     ResultCode(61, ErrorModule::HTTP, ErrorSummary::InvalidState, ErrorLevel::Permanent);
 
+static std::pair<std::string, std::string> SplitUrl(const std::string& url) {
```
Comment (GPUCode) [inline src/core/hle/service/http/http_c.cpp:58]:
```suggestion
static std::pair<std::string_view, std::string_view> SplitUrl(std::string_view url) {
```
The function operates strictly on substrings of the input so it does not need to make copies

File: src/core/hle/service/http/http_c.cpp [hunk 7]
```diff
@@ -55,11 +55,31 @@ const ResultCode ERROR_WRONG_CERT_HANDLE = // 0xD8A0A0C9
...
+    return std::make_pair(host, path);
+}
+
 void Context::MakeRequest() {
     ASSERT(state == RequestState::NotStarted);
 
 #ifdef ENABLE_WEB_SERVICE
-    std::unique_ptr<httplib::Client> client = std::make_unique<httplib::Client>(url.c_str());
+    const auto& [host, path] = SplitUrl(url.c_str());
+    std::unique_ptr<httplib::Client> client = std::make_unique<httplib::Client>(host);
```
Comment (GPUCode) [inline src/core/hle/service/http/http_c.cpp:82]:
```suggestion
    const auto& [host, path] = SplitUrl(url);
    const auto client = std::make_unique<httplib::Client>(host);
```

**Review by GPUCode — APPROVED** [review]

**Comment by GPUCode** [comment]
I realised httplib does not accept string_view in its constructor so feel free to ignore these comments. string_view is the ideal but would make the code more complex this this case

File: src/core/hle/service/http/http_c.cpp [hunk 8]
```diff
@@ -549,40 +486,55 @@ void HTTP_C::AddPostDataAscii(Kernel::HLERequestContext& ctx) {
...
     }
 
-    if (session_data->current_http_context != context_handle) {
-        LOG_ERROR(Service_HTTP,
-                  "Tried to add set client cert on a mismatched session input context={} session "
-                  "context={}",
-                  context_handle, *session_data->current_http_context);
-        IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
-        rb.Push(ERROR_STATE_ERROR);
+    const u32 response_code = itr->second.response.status;
```
Comment (neobrain) [inline src/core/hle/service/http/http_c.cpp:522]: This and other lines in this file won't compile with `!ENABLE_WEB_SERVICE`.

**Pull request merged**